File: Python/Problem1.py
# Problem 1
# Title:    Multiples of 3 or 5
# Author:   Paul Kim
# Date:     6/9/2017

# Description
# If we list all the natural numbers below 10 that are multiples of 3 OR 5, we
# get 3, 5, 6, and 9. The sum of these multiples is 23.
# Find the sum of all the multiples of 3 or 5 below 1000.

#####UNOPTIMIZED METHOD#####
# Modulo three or five multiple method
def three_five_mult( counter ):
    if counter % 3 == 0:
        return counter
    elif counter % 5 == 0:
        return counter
    else:
        return -1
# Summation method that utilizes three or five Modulo method
def three_five_mult_sum( max ):
    sum = 0
    for i in range(1,max):
        if not(three_five_mult(i) == -1):
            sum += i
    return sum
#####UNOPTIMIZED METHOD#####



#####OPTIMIZED METHOD#####
def three_five_mult_sumOpt( max ):
    sum = 0
    isThree = 0
    isFive = 0
    either = False
    for i in range(1, max):
        either = False
        isThree += 1
        isFive += 1
        if isThree == 3:
            either = True
            isThree = 0
        if isFive == 5:
            either = True
            isFive = 0

        if either:
            sum += i

    return sum
#####OPTIMIZED METHOD#####



#####MAIN METHOD#####
upperBound = 1000
format = "sum of all whole numbers below %d is: %d."

# unoptimized method call
# sum = three_five_mult_sum(upperBound)
# print(format % (upperBound, sum))

# optimized method call
sum = three_five_mult_sumOpt(upperBound)
print(format % (upperBound, sum))
# sum of all whole numbers below 1000 is: 233168.

File: Java/Problem1/Problem1.java
import java.math.BigInteger;
/**
 * Problem 1
 * Title:    Multiples of 3 or 5
 * Author:   Paul Kim
 * Date:     6/9/2017
 *
 * Description
 * If we list all the natural numbers below 10 that are multiples of 3 OR 5, we
 * get 3, 5, 6, and 9. The sum of these multiples is 23.
 * Find the sum of all the multiples of 3 or 5 below 1000.
 */
public class Problem1
{
  public static void main(String[] args)
  {
    String format = "sum of all 3 or 5 multiples below %d is: %s.\n";

    int upperBound = 1000;
    BigInteger sum = threeFiveSum(upperBound);

    // sum of all 3 or 5 multiples below 1000 is: 233168.
    System.out.printf(format, upperBound, sum.toString());
  }

  private static BigInteger threeFiveSum( int upperBound )
  {
    BigInteger sum = BigInteger.ZERO;
    int i, isThree = 0, isFive = 0;
    boolean isEither;
    for(i = 1; i < upperBound; ++i )
    {
      isEither = false;
      if( ++isThree == 3 )
      {
        isEither = true;
        isThree = 0;
      }
      if( ++isFive == 5 )
      {
        isEither = true;
        isFive = 0;
      }

      if( isEither )
        sum = sum.add(BigInteger.valueOf(i));
    }

    return sum;
  }
}
